### PR TITLE
Remove psutil as a dependency

### LINKED ISSUE #11254: Remove psutil as a dependency
### What should we add?

Right now we have `psutil` as a dependency which is used to get the number of *physical* CPUs and also the available amount of memory at the local system level. The physical CPU count is used to determine the max default number of parallel processes in `parallel_map` and the memory information is used to determine statevector size limits for `BasicAer` and also is printed in the jupyter version table magic. However, relying on an external lib for just these small uses seems a bit excessive primarily because psutil doesn't have precompiled wheels available for all our supported platforms which requires users on those platforms to have the appropriate compiler and library headers to build it from source to use qiskit. But additionally the checking we're doing via psutil in these cases isn't actually sound. These checks only look at what's reported from the OS for the entire system, but don't actually reflect any process limits that might be in place (this is documented in the psutil docs) so we shouldn't be making task scheduling or job limit decisions based on this information. 

As discussed in #10868 we're using the physical core count in `parallel_map()` to try and maximize performance because in the case of systems with SMT, scheduling jobs on each logical core can have lower performance. I think a reasonable approach is to default `CPU_COUNT` to be `int(len(os.sched_getaffinity(0)) / 2)`, as most (but obviously not all) modern systems have 2 way SMT. This obviously isn't as robust a detection solution to avoid using logical cores, but I think as a default it's reasonable especially as for any users that want different parallel dispatch behavior there is the `QISKIT_NUM_PROCS` env variable or the `num_processes` setting in the qiskit config file (https://qiskit.org/documentation/configuration.html).

The memory usage we can just remove the limits from basic aer and set the qubit count to 24 as that's ~270MB and realistically is always the limit for users. If there isn't sufficient memory available for a 24 qubit statevector for some reason, then numpy will error on allocation anyway. The only use case not covered is the jupyter magic, but I think we can just drop the memory reported from that as it's not super critical functionality.

### PR DESCRIPTION
**Pull request by FabianBrings**
# Summary

Removes any uses of psutil and the corresponding dependency

# Details and comments

As described in issue #11254, psutil is used to detect the number of cores and memory to optimize resource allocation. The actual usefulness of those information is however questionable, while adding additional bloat. Therefore any references to the module were modified, the unit tests fixed and the dependency removed.

Fixes #11254

### CODE REVIEW

**Comment by FabianBrings** [comment]
The test that fails is "test_handle_measurement" while performing SabreSwaps. I can reproduce the error, by setting QISKIT_NUM_PROCS=1. SabreSwap uses CPU_COUNT to determine the number of runs while evaluating the heuristic to optimize the circuit. The new method of finding the cpu count probably yields a different, platform-dependent result and therefore the heuristic returns another result. This raises the question, how this test case should be handled. If the reference circuit is not reached depending on system parameters, it is not a sensible test case, but hard overwriting the cpu count with two, just to make this case work seems equally nonsensical. How would you like to proceed?

**Comment by mtreinish** [comment]
It's a bug in the test, we should be setting a fixed number of trials to ensure the test is deterministic. So I'd suggest ust hardcoding the trial count in the test.

File: qiskit/utils/multiprocessing.py [hunk 1]
```diff
@@ -27,13 +26,18 @@ def local_hardware_info():
     Returns:
         dict: The hardware information.
     """
+
+    if hasattr(os, "sched_getaffinity"):
+        num_cpus = int(len(os.sched_getaffinity(0)) / 2) or 1
+    else:
+        num_cpus = 1
```
Comment (mtreinish) [inline qiskit/utils/multiprocessing.py:35]: I think in this case we can do:

```suggestion
        num_cpus = os.cpu_count()
        if num_cpus is None:
            num_cpus = 1
        else:
            num_cpus = int(num_cpus / 2) or 1
```

File: test/python/test_util.py [hunk 2]
```diff
@@ -21,11 +22,11 @@
 class TestUtil(QiskitTestCase):
     """Tests for qiskit/_util.py"""
 
+    @skipIf(not hasattr(os, "sched_getaffinity"), "ched_getaffinity is only available on linux")
```
Comment (mtreinish) [inline test/python/test_util.py]: With the changes suggested above is this needed?

File: releasenotes/notes/psutil-dependancy-removed-bf5366f516d92378.yaml [hunk 3]
```diff
@@ -0,0 +1,4 @@
+---
+upgrade:
+  - |
+    Removed psutil for detecting number of cpus and memory. the information provided does not add sufficient value to justify the additional dependencies and overhead.
```
Comment (mtreinish) [inline releasenotes/notes/psutil-dependancy-removed-bf5366f516d92378.yaml]:
```suggestion
    The dependency on `psutil <https://pypi.org/project/psutil/>`__ has been removed. The
    psutil library was only used for detecting the number of physical CPUs and total system
    memory, however this information provided does not add sufficient value to justify the
    additional dependencies and overhead so it has been removed. This does mean that the
    default number of processes used by :func:`.parallel_map` and functions that internally
    can use :func:`.parallel_map` such as :func:`.transpile` and :meth:`.PassManager.run` may
    use more or less parallel processes than in previous releases. If you'd like to adjust the
    number of processes used you can use the ``QISKIT_NUM_PROCS`` environment variable or
    ``num_processes`` field in a user configuration file (see the
    `local configuration guide <https://docs.quantum.ibm.com/start/configure-qiskit-local>`__
    for more details) if you need to adjust the number of processes that Qiskit potentially uses.
```

**Review by mtreinish — COMMENTED** [review]
Overall this LGTM, thanks for tackling this. I just had a few inline comments.

File: test/python/test_util.py (hunk 2 above)
Comment (FabianBrings) [reply test/python/test_util.py]: "Mock" was giving an error about sched_getaffinity not existing, so I assumed, that it still tries to access the method, to setup the mock-method. This is why I added the skip-decorator.

**Review by mtreinish — DISMISSED** [review]
Thanks for opening the PR. I just pushed up some small commits to update the tests and apply my inline suggestions to get this unblocked

**Review by mtreinish — DISMISSED** [review]

**Review by mtreinish — APPROVED** [review]

**Comment by FabianBrings** [comment]
Thanks a lot for approving my PR and for helping out with the small fixes!

**Pull request merged**